WIP and a proposal for updating the styles doc. (#498)
Polymer/lit-element: docs/_guide/styles.md

slug: styles
---

> 

Reviewer: I'm wanting an intro section with a more general name. Like:

Overview

LitElement provides flexibility in how you style your components, and ensures interoperability through style encapsulation using Shadow DOM, the browser's built-in way of encapsulating native UI components.

Why use Shadow DOM? Shadow DOM scopes CSS so that...

## Shadow DOM

> LitElement uses [Shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM) by default for DOM and style encapsulation. Shadow DOM scopes CSS so that styles defind in a shadow root only apply to DOM inside the shadow root and do not leak to outside DOM. Shadow roots are also isolated from styles defined outside the shadow root, whether in the main page or a outer shadow root.

Reviewer: ```suggestion
LitElement uses [Shadow DOM](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_shadow_DOM) by default for DOM and style encapsulation. Shadow DOM scopes CSS so that styles defined in a shadow root only apply to DOM inside the shadow root and do not leak to outside DOM. Shadow roots are also isolated from styles defined outside the shadow root, whether in the main page or an outer shadow root.
```

**This guide applies only if you use the default (shadow DOM) render root.** If you modify your element's render root to render into the main DOM tree instead of a shadow root, these instructions won't apply.
{.alert-info}

**If you're using the Shady CSS polyfill, be aware that it has some limitations.** See the [Shady CSS README](https://github.com/webcomponents/shadycss/blob/master/README.md#limitations) for more information.
{.alert-info}

{::options toc_levels="1..3" /}
* ToC
{:toc}

> ### Styling shadow DOM

Reviewer: ```suggestion
### Styling your custom elements
```

> Styles can be added to a shadow root in three different ways:

Reviewer: ```suggestion
Styles can be added to a custom element in three different ways:
```

* Elements in the shadow root
* Slotted elements: usually light DOM children that are projected into a `<slot>` element.

## Static styles

LitElements lets you define styles that apply to all instances with a static `styles` property on your element class.

The value of the `styles` property must be created with the `css` template literal tag provided by the `lit-element` module. `styles` can either be a single value, or an array.

An example of a single style:
```js
import {LitElement, css} from 'lit-element';

class MyElement extends LitElement {
  static styles = css`
    :host {
      display: block;
    }
  `;
}
```

Multiple values are useful for inheritance:

```js
class MySubElement extends MyElement {
  static styles = [
    super.styles,
    css`
      :host([.important]) {
        color: red;
      }
    `
  ];
}
```

Here the subclass is explicitly inheriting the superclass styles, then adding its own.

Static styles are the reccomended way to style LitElements because on browsers that support it, they generate and use a new standard called _[Constuctible Stylesheets](https://wicg.github.io/construct-stylesheets/)_, with a fallback for browsers that don't.

> Constuctible Stylesheets allow LitElement to parse styles exactly once and reuse the resulting Stylesheet object for maximum efficiency.

Reviewer 2: Const*r*uctable

### Limitations

> Static styles apply to all instances of an element. Any expressions in the style text are evaluated and included once, then reused for all instances. If you need to vary styles per-instance we reccomend using CSS custom properties, or inline styles.

Reviewer 2: Recommend (one c)

For security reasons, the only types of values that can be interpolated into static styles are values returned by the `cssLiteral` and `unsafeCSS` template tags.

```js
const mainColor = cssLiteral`red`;

class MyElement extends LitElement {
  static styles = css`
    :host {
      display: block;
      color: ${mainColor}
    }
  `;
}
```

## Inline styles

> You can also style a shadow root with inline styles right in your element template. We still reccomend static styles, but in some cases you may want to vary the CSS per-element. One way to do this is with bindings in `<style>` elements. It's important to note that this will not work with shadow DOM polyfills like ShadyCSS.

Reviewer 2: Recommend again

```js
import {LitElement, property} from 'lit-element';

class MyElement extends LitElement {
  @property() mainColor = 'blue';
  render() {
    return html`
      <style>
        :host {
          color: ${this.mainColor};
        }
      </style>
    `;
  }
}
```

> We strongly reccomend static styles, CSS custom properties, or [lit-html's `classMap` or `styleMap` directives](TODO) if you're stying non-host shadow root contents.

Reviewer 3: what does non-host shadow root contents mean? could you give an example? I'm not sure we need to cover it

## External stylesheets

You can load an external stylesheet into a shadow root with a `<link>` element:

```js
import {LitElement} from 'lit-element';

class MyElement extends LitElement {
  render() {
    return html`
      <link rel="stylesheet" href="./styles.css">
    `;
  }
}
```

This can be a good way to load CSS generated from tools like SASS/LESS.

There are some important caveats though:

* External styles can cause a flash-of-unstyled-content (FOUC) while they load.

> * The URL in the `href` attribute is relative to the _main document_, making this technique mostly useful for application elements where asset URLs are well known, and not for reusable elements published publically.

Reviewer 3: * As we are not documenting how to build apps, (covered in PWA starter kit) should this content be here or somewhere else? @jbingham @graynorton 
* If it belongs here, someone is bound to ask how to get the URL the the element was imported from (can't remember how to do it, import meta something something..) should we cover that somewhere?
